### PR TITLE
Implemented Floor

### LINKED ISSUE #3255: implement intrinsic `floor`
Implement in `intrinsic_function_registry.h`.
Hint: Take idea from `floor` implementation in `comptime_eval.h`, `ast_common_visitor.h` and `src/runtime/pure/lfortran_intrinsic_trig.f90` and finally remove them.

### PR DESCRIPTION
**Pull request by parth121101**
fix #3255

### CODE REVIEW

**Comment by parth121101** [comment]
```
$ cat a.f90
program main
    real :: x = 5.8
    print *, floor(x)
end program
 lfortran a.f90
1
2
3
4
5
Internal Compiler Error: Unhandled exception

```
I am trying to test my dummy file for the floor implementation. I am facing these issue. Can anyone help me out to resolve this.

**Comment by Pranavchiku** [comment]
This happens because of incorrect types at multiple places, apply the following diff:

```diff
diff --git a/src/libasr/pass/intrinsic_function_registry.h b/src/libasr/pass/intrinsic_function_registry.h
index c528060f5..d386a6871 100644
--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -2081,17 +2081,16 @@ namespace Floor {
     static inline ASR::asr_t* create_Floor(Allocator& al, const Location& loc,
             Vec<ASR::expr_t*>& args,
             const std::function<void (const std::string &, const Location &)> err) {
-        ASR::ttype_t* return_type = expr_type(args[0]);
+        ASR::ttype_t* return_type = TYPE(ASR::make_Integer_t(al, loc, 4));
         if (!(args.size() == 1 || args.size() == 2)) {
             err("Intrinsic `Floor` function accepts exactly 1 or 2 arguments", loc);
-        } else if (!ASRUtils::is_real(*return_type)) {
+        } else if (!ASRUtils::is_real(*ASRUtils::expr_type(args[0]))) {
             err("Argument of the `Floor` function must be Real", args[0]->base.loc);
         }
         Vec<ASR::expr_t *> m_args; m_args.reserve(al, 1);
         m_args.push_back(al, args[0]);
         if ( args[1] != nullptr ) {
             int kind = -1;
-            std::cout<<"in bere"<<std::endl;
             if (!ASR::is_a<ASR::Integer_t>(*expr_type(args[1])) ||
                     !extract_value(args[1], kind)) {
                 err("`kind` argument of the `Floor` function must be a "
@@ -2101,14 +2100,14 @@ namespace Floor {
         }
         ASR::expr_t *m_value = nullptr;
 
-        if (all_args_evaluated(args)) {
+        if (all_args_evaluated(m_args)) {
             Vec<ASR::expr_t*> arg_values; arg_values.reserve(al, 1);
             arg_values.push_back(al, expr_value(args[0]));
-            m_value = eval_Floor(al, loc, expr_type(args[0]), arg_values);
+            m_value = eval_Floor(al, loc, return_type, arg_values);
         }
         return ASR::make_IntrinsicScalarFunction_t(al, loc,
             static_cast<int64_t>(IntrinsicScalarFunctions::Floor),
-            m_args.p, m_args.n, 0, ASRUtils::expr_type(args[0]), m_value);
+            m_args.p, m_args.n, 0, return_type, m_value);
     }
 
     static inline ASR::expr_t* instantiate_Floor(Allocator &al, const Location &loc,
@@ -2125,35 +2124,29 @@ namespace Floor {
         *   r = int(x)
         * }
         */
-       std::cout<<"1"<<"\n";
-        ASR::expr_t *one = i(1, arg_types[0]);
-       std::cout<<"2"<<"\n";
+        ASR::expr_t *one = i(1, return_type);
+        ASR::expr_t *cast = ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[0], ASR::cast_kindType::RealToInteger, return_type, nullptr));
 
-        body.push_back(al, b.If(fLt(args[0], i(0, arg_types[0])), {
+        body.push_back(al, b.If(fLt(args[0], f(0, arg_types[0])), {
             b.Assignment(
                 result,
                 i_tSub(
-                    ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[0], ASR::cast_kindType::RealToInteger, return_type, nullptr)),
+                    cast,
                     one,
-                    arg_types[0]
+                    return_type
                 )
             )
         }, {
             b.Assignment(
                 result,
-                ASRUtils::EXPR(ASR::make_Cast_t(al, loc, args[0], ASR::cast_kindType::RealToInteger, return_type, nullptr))
+                cast
             )
         }));
-       std::cout<<"3"<<"\n";
 
         ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
             body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
-       std::cout<<"4"<<"\n";
         scope->add_symbol(fn_name, f_sym);
-       std::cout<<"5"<<"\n";
-
         return b.Call(f_sym, new_args, return_type, nullptr);
-
     }
 
 } // namespace Floor
```

**Comment by Pranavchiku** [comment]
Here is the test file for kind = 4, 8. You may add tests for kind = 1, 2 if you wish though it is not a compulsion from my end

```fortran
program main
    real :: x = 5.8
    integer(kind = 4) :: res_4
    integer(kind = 8) :: res_8
    
    res_4 = floor(x)
    print *, res_4
    if (res_4 /= 5) error stop

    res_8 = floor(x, 8)
    print *, res_8
    if (res_8 /= 5) error stop

    res_4 = floor(x, 4)
    print *, res_4
    if (res_4 /= 5) error stop

    res_4 = floor(5.8)
    print *, res_4
    if (res_4 /= 5) error stop

    res_8 = floor(5.8, 8)
    print *, res_8
    if (res_8 /= 5) error stop

    res_4 = floor(0.0)
    print *, res_4
    if (res_4 /= 0) error stop

    res_8 = floor(0.0, 8)
    print *, res_8
    if (res_8 /= 0) error stop

    res_4 = floor(-412.124)
    print *, res_4
    if (res_4 /= -413) error stop

    res_8 = floor(-412.124, 8)
    print *, res_8
    if (res_8 /= -413) error stop

end program
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
commented above.

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 1]
```diff
@@ -170,6 +175,8 @@ def add_arg_type_src(func_name):
     src += 3 * indent + f'ASRUtils::require_impl(false, "Unexpected number of args, {func_name} takes {no_of_args_msg} arguments, found " + std::to_string(x.n_args), x.base.base.loc, diagnostics);\n'
     src += 2 * indent + "}\n"
 
+    
+
```
Comment (Pranavchiku) [inline src/libasr/intrinsic_func_registry_util_gen.py:179]: Please remove these spaces.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -2062,6 +2064,93 @@ namespace Anint {
...
+                    one,
+                    return_type
+                )
+            )
+        }, {
+            b.Assignment(
+                result,
+                cast
+            )
+        }));
```
Comment (Pranavchiku) [inline src/libasr/pass/intrinsic_function_registry.h:2144]: collapse these into one line

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Please do a clean build and then update references following the commands:

```
rm -rf ./tests/reference/*
./run_tests.py -u
```

**Review by Pranavchiku — APPROVED** [review]
Looks good to me. I will wait for another pair of eyes to review this. Thanks!

**Review by HarshitaKalani — COMMENTED** [review]
Looks good to me too!

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
The CI failure was due to https://github.com/lfortran/lfortran/issues/3282. I restarted it.

**Pull request merged**